pkg/jobs/jobs.go

```diff
@@ -750,6 +752,78 @@ func HasJobNotFoundError(err error) bool {
 	return errors.HasType(err, (*JobNotFoundError)(nil))
 }
 
+func (j *Job) loadJobPayloadAndProgress(
+	ctx context.Context, st *cluster.Settings, txn isql.Txn,
+) (*jobspb.Payload, *jobspb.Progress, error) {
+	if txn == nil {
+		return nil, nil, errors.New("cannot load job payload and progress with a nil txn")
+	}
+
+	payload := &jobspb.Payload{}
+	progress := &jobspb.Progress{}
+	if st.Version.IsActive(ctx, clusterversion.V23_1JobInfoTableIsBackfilled) {
+		infoStorage := j.InfoStorage(txn)
+
+		payloadBytes, exists, err := infoStorage.GetLegacyPayload(ctx)
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, "failed to get payload for job %d", j.ID())
+		}
+		if !exists {
+			return nil, nil, errors.Wrap(&JobNotFoundError{jobID: j.ID()}, "job not found in system.job_info")
+		}
+		if err := protoutil.Unmarshal(payloadBytes, payload); err != nil {
+			return nil, nil, err
+		}
+
+		progressBytes, exists, err := infoStorage.GetLegacyProgress(ctx)
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, "failed to get payload for job %d", j.ID())
+		}
+		if !exists {
```

Comment:
do we intend on making progress optional at some point?

Reply:
_maybe_. On job creation we expect the job record to have a valid, non-null progress and payload so unless the user manually deletes the progress for a job from `system.job_info` we will always have an entry. That said `SHOW JOBS` goes out of its way to handle the absence of a progress row for a job. So I dunno we should decide if having progress is mandatory or not and then apply that through the system.